test/test-remote/testutils/deployment.ts

```diff
@@ -0,0 +1,283 @@
+type State = ReturnType<typeof rootReducer>;
+
+function* runInformers(
+  kubeConfig: KubeConfig
+): Generator<
+  ChannelTakeEffect<void | unknown> | PutEffect | CancelledEffect,
+  void,
+  Action
+> {
+  log.info(`Starting pod informers`);
+
+  const clusterChannel = eventChannel(channel => {
+    const unsubscribes = [
+      DaemonSet.startInformer(kubeConfig, channel),
+      Deployment.startInformer(kubeConfig, channel),
+      StatefulSet.startInformer(kubeConfig, channel)
+    ];
+
+    // return the unsubscribe function for eventChannel. This will be called when the channel
+    // is closed.
+    return () => {
+      log.info(`Closing clusterChannel and shutting down pod informers`);
+      unsubscribes.forEach(fn => fn());
+    };
+  });
+
+  try {
+    while (true) {
+      const event = yield take(clusterChannel);
+      yield put(event);
+    }
+  } catch (e) {
+    log.error(e);
+  } finally {
+    // If task cancelled, close the channel, unsubscribing the informers
+    if (yield cancelled()) {
+      clusterChannel.close();
+    }
+  }
+}
+
+function* blockUntilCacheHydrated(): Generator<
+  SelectEffect | CallEffect,
+  void,
+  CombinedState<State>
+> {
+  while (true) {
+    const { kubernetes }: State = yield select();
+    const {
+      DaemonSets,
+      Deployments,
+      StatefulSets
+    } = kubernetes.cluster;
+
+    if (
+      DaemonSets.loaded &&
+      Deployments.loaded &&
+      StatefulSets.loaded
+    ) {
+      log.info(`kubernetes cache is hydrated`);
+      break;
+    }
+    log.info(`waiting for the kubernetes cache to hydrate...`);
+    yield delay(3000);
+  }
+}
+
+//eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+function* waitForReady(config: {
+  originalDaemonSets: DaemonSet[],
+  originalDeployments: Deployment[],
+  originalStatefulSets: StatefulSet[];
+}): Generator<Effect, void, State>  {
+  while (true) {
+    const state: State = yield select();
+    const { DaemonSets, Deployments, StatefulSets } = state.kubernetes.cluster;
+    const matchingDaemonSets = config.originalDaemonSets.map(o => getMatch(o, DaemonSets.resources));
+    const matchingDeployments = config.originalDeployments.map(o => getMatch(o, Deployments.resources));
+    const matchingStatefulSets = config.originalStatefulSets.map(o => getMatch(o, StatefulSets.resources));
+
+    // Check if any of the items are "active" (still deploying), log the resulting messages if so
+    const activeMessages = activeDaemonsets(matchingDaemonSets)
+      .concat(activeDeployments(matchingDeployments))
+      .concat(activeStatefulsets(matchingStatefulSets));
+    if (activeMessages.length == 0) {
+      log.info(`All pod resources are Ready, exiting wait`);
+      break;
+    }
+
+    log.info(`Waiting on ${activeMessages.length} active pod resources:\n- ${activeMessages.join("\n- ")}`);
+    yield delay(7000);
+  }
+}
+
+function getMatch<T extends K8sResource>(needle: T, haystack: T[]): T {
```

Comment:
I love when someone chooses the variable names `needle` and `haystack`. I also do that, it's one of those many examples of super expressive variable names -- communicates so much of the intent in so little time :)